### PR TITLE
Added deepcopy implementation.

### PR DESCRIPTION
**Pull request by hazemessamm**
## Proposed changes

Added `deepcopy` implementation to create more than one copy for the same model.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/mlx/nn/layers/base.py [hunk 1]
```diff
@@ -90,6 +91,34 @@ def __getattr__(self, key: str):
...
+        cls = self.__class__
+        instance_copy = cls.__new__(cls)
+        memo[id(self)] = instance_copy
+
+        for k, v in self.items():
+            if isinstance(v, mx.array):
+                obj = mx.array(v)
+            elif isinstance(v, list):
+                obj = [
+                    mx.array(elem) if isinstance(elem, mx.array) else deepcopy(elem)
```
Comment (hazemessamm) [inline python/mlx/nn/layers/base.py:104]: This solution might not be the best but I did not find a method that copies `mx.array` so I just copied it by calling `mx.array(old_mx_array)` to create a new array.

**Comment by angeloskath** [comment]
@hazemessamm out of curiosity why would we want a deep copy?

In general weights are not updated in place so keeping a copy of the parameters dictionary means you have a copy of the weights. Even if someone changes them. Basically what does this PR make possible with the deep copy except for the ability to do something like

```
linear.weight *= 2
```

Keep in mind that doing

```
old_weights = linear.parameters()
linear.weight = linear.weight * 2
```

does not affect `old_weights` at all.

**Comment by hazemessamm** [comment]
Thanks for the response and explanatikn I did not know that, I was working on an Exponential Moving Average module where I wanted to take a deep copy of the model so I can update it and avoid changing anything in the original model. Do you think is there a way to do that without deep copying the model?

**Comment by angeloskath** [comment]
Oh yeah exponential moving average would be extremely easy to do in MLX. Literally the following code:

```python
def ema_update(ema_params, model, alpha=0.9):
    return tree_map(lambda a, b: a * alpha + (1-alpha) * b, ema_params, model.parameters())

ema_parameters = model.parameters()
for batch in iterate_batches(data):
    loss, grad = loss_and_grad(model, batch)
    optimizer.update(model, grad)
    del grad  # to save some memory cause we don't need it any more
    ema_parameters = ema_update(ema_parameters, model)
    mx.eval(model.parameters(), ema_parameters, optimizer.state)
```

To use the ema weights afterwards one needs simply to do `model.update(ema_parameters)` and then use the model normally.

I will close this PR since I think a deep copy is almost never needed and for sure not for EMA :-) . Feel free to reopen if you think there is something that needs a deep copy.

**Pull request closed without merging**